### PR TITLE
wireless: add frequency-list in station mode (jsc#PED-8715)

### PR DESCRIPTION
**Pull request by cfconrad**
* introduce new `WIRELESS_FREQUENCY_LIST` and `<frequency-list>`
  * It is possible to specify a frequency directly or use one of the following aliases `2.5GHz`, `5GHz`, `6GHz`.
* add `channel/frequency` info into `wicked ifstatus --verbose wlanX`
* add `frequency` to scan results

### Example wicked ifstatus output:
```bash
tw-1:~ # wicked ifstatus --verbose  wlan1
wlan1           up
      link:     #6, state up, mtu 1500
      type:     wireless, state established, ssid MySSID Muha, WPA2-PSK+WPA-PSK
      wireless: bssid 02:00:00:00:00:00, signal -30, channel 6 (2437MHz), duration 680s
      control:  none

...
```

### Manpage output:
Screenshot from 2024-07-16 11-18-39

### CODE REVIEW

File: src/wireless.c [hunk 1]
```diff
@@ -407,23 +407,128 @@ ni_wireless_scan_add_bss(ni_wireless_scan_t *scan, ni_wireless_bss_t *bss)
...
+	} else if (frequency == 2484) {
 		return 14;
 
 	} else {
 		return (frequency - 2407) / 5;
 	}
 }
 
+ni_bool_t
+ni_wireless_band2freq_list(ni_uint_array_t *list, const char *value)
```
Comment (mtomaschewski) [inline src/wireless.c:431]:
```suggestion
ni_wireless_band2freq_list(ni_uint_array_t *array, const char *value)
```

File: src/wireless.c [hunk 2]
```diff
@@ -407,23 +407,128 @@ ni_wireless_scan_add_bss(ni_wireless_scan_t *scan, ni_wireless_bss_t *bss)
...
+{
+	static const unsigned int chan_2_4ghz[] = {
```
Comment (mtomaschewski) [inline src/wireless.c:433]: This aren't channels but frequencies:

```suggestion
	static const unsigned int freq_set_2_4ghz[] = {
```

File: src/wireless.c [hunk 3]
```diff
@@ -407,23 +407,128 @@ ni_wireless_scan_add_bss(ni_wireless_scan_t *scan, ni_wireless_bss_t *bss)
...
+		2412, 2417, 2422, 2427, 2432, 2437, 2442,
+		2447, 2452, 2457, 2462, 2467, 2472, 2484,
+		0
+	};
+	static const unsigned int chan_5ghz[] = {
```
Comment (mtomaschewski) [inline src/wireless.c:438]:
```suggestion
	static const unsigned int freq_set_5ghz[] = {
```

File: src/wireless.c [hunk 4]
```diff
@@ -407,23 +407,128 @@ ni_wireless_scan_add_bss(ni_wireless_scan_t *scan, ni_wireless_bss_t *bss)
...
+		5935, 5955, 5975, 5995, 6015, 6035, 6055, 6075, 6095, 6115, 6135, 6155, 6175,
+		6195, 6215, 6235, 6255, 6275, 6295, 6315, 6335, 6355, 6375, 6395, 6415, 6435,
+		6455, 6475, 6495, 6515, 6535, 6555, 6575, 6595, 6615, 6635, 6655, 6675, 6695,
+		6715, 6735, 6755, 6775, 6795, 6815, 6835, 6855, 6875, 6895, 6915, 6935, 6955,
+		6975, 6995, 7015, 7035, 7055, 7075, 7095, 7115,
+		0
+	};
+
+	typedef struct {
+		const char *name;
```
Comment (mtomaschewski) [inline src/wireless.c:456]:
```suggestion
		ni_wireless_freqency_set_t	set;
```

File: src/wireless.c [hunk 5]
```diff
@@ -407,23 +407,128 @@ ni_wireless_scan_add_bss(ni_wireless_scan_t *scan, ni_wireless_bss_t *bss)
...
+		const unsigned int *list;
```
Comment (mtomaschewski) [inline src/wireless.c:457]:
```suggestion
		const unsigned int *	array;
```

File: src/wireless.c [hunk 6]
```diff
@@ -407,23 +407,128 @@ ni_wireless_scan_add_bss(ni_wireless_scan_t *scan, ni_wireless_bss_t *bss)
...
+	} ni_wireless_band2freq_t;
+
+	static const ni_wireless_band2freq_t band2freq[] = {
+		{ "2.4GHz", chan_2_4ghz},
```
Comment (mtomaschewski) [inline src/wireless.c:461]:
```suggestion
		{ NI_WIRELESS_FREQUENCY_SET_2_4GHz,	freq_set_2_4ghz	},
```

File: src/wireless.c [hunk 7]
```diff
@@ -407,23 +407,128 @@ ni_wireless_scan_add_bss(ni_wireless_scan_t *scan, ni_wireless_bss_t *bss)
...
+	if (ni_parse_ulong(value, &tmp, 10) == 0) {
+		if (tmp < UINT_MAX)
+			return !list || ni_uint_array_append(list, (uint) tmp);
+	}
+
+	return ni_wireless_band2freq_list(list, value);
+}
+
+ni_bool_t
+ni_wireless_frequency_list_expand_csv(ni_uint_array_t *expanded_freq, const char *value,
```
Comment (mtomaschewski) [inline src/wireless.c:501]:
```suggestion
ni_wireless_frequency_list_parse_string(const char *value, ni_uint_array_t *expanded_freq,
```

**Review by mtomaschewski — CHANGES_REQUESTED** [review]
This pull request still needs some work, love and tests (error handling).

File: client/suse/compat-suse.c [hunk 8]
```diff
@@ -4058,6 +4058,23 @@ try_add_wireless_net(const ni_sysconfig_t *sc, ni_netdev_t *dev, const char *suf
...
+		ni_stringbuf_t tmp = NI_STRINGBUF_INIT_DYNAMIC;
+
+		if (!ni_wireless_frequency_list_expand_csv(NULL, var->value, &net->frequency_list, &errors)) {
+			ni_error("ifcfg-%s: wrong WIRELESS_FREQUENCY_LIST%s value '%s'",
+					dev->name, suffix,
+					ni_stringbuf_join(&tmp, &errors, ","));
+
+			ni_string_array_destroy(&errors);
+			ni_stringbuf_destroy(&tmp);
+			return FALSE;
```
Comment (cfconrad) [inline client/suse/compat-suse.c:4074]: Do not invalidate the full config, it is enough to throw a warning for wrong values.

File: schema/wireless.xml [hunk 9]
```diff
@@ -16,6 +16,7 @@
     <mode type="builtin-wireless-mode"/>
     <access-point type="ethernet-address" description="Address of AP to connect to"/>
     <channel type="uint32"/>
+    <frequency-list type="string"/>
     <fragment-size type="uint32"/>
     <frequency type="double"/>
```
[the remark this answers is not in the document]
Comment (cfconrad) [reply schema/wireless.xml:21]: I didn't wanted to touch this. But If we do so, then I would simply remove this value. As it isn't used elsewhere.

File: client/suse/compat-suse.c [hunk 10]
```diff
@@ -4058,6 +4058,22 @@ try_add_wireless_net(const ni_sysconfig_t *sc, ni_netdev_t *dev, const char *suf
...
+	if (net->mode == NI_WIRELESS_MODE_MANAGED &&
+	    (var = __find_indexed_variable(sc, "WIRELESS_FREQUENCY_LIST", suffix))) {
+
+		ni_string_array_t errors = NI_STRING_ARRAY_INIT;
+		ni_stringbuf_t tmp = NI_STRINGBUF_INIT_DYNAMIC;
+
+		if (!ni_wireless_frequency_list_parse_string(var->value, &net->frequency_list, &errors)) {
+			ni_warn("ifcfg-%s: invalid WIRELESS_FREQUENCY_LIST%s value(s) '%s'",
+					dev->name, suffix,
+					ni_stringbuf_join(&tmp, &errors, ","));
```
Comment (mtomaschewski) [inline client/suse/compat-suse.c:4070]: I'd use a space " " here as well:
```suggestion
					ni_stringbuf_join(&tmp, &errors, " "));
```

File: src/dbus-objects/wireless.c [hunk 11]
```diff
@@ -391,6 +391,23 @@ ni_objectmodel_get_wireless_request_net(const char *ifname, ni_wireless_network_
...
 
+	if (ni_dbus_dict_get_string(var, "frequency-list", &string)) {
+		ni_string_array_t errors = NI_STRING_ARRAY_INIT;
+		ni_stringbuf_t tmp = NI_STRINGBUF_INIT_DYNAMIC;
+
+		if (!ni_wireless_frequency_list_parse_string(string, &net->frequency_list, &errors)) {
+			if (!dbus_error_is_set(error))
+				dbus_set_error(error, DBUS_ERROR_INVALID_ARGS,
+						"%s: invalid frequency-list value %s",
+						ni_stringbuf_join(&tmp, &errors, ","),
```
Comment (mtomaschewski) [inline src/dbus-objects/wireless.c:402]:
```suggestion
						ni_stringbuf_join(&tmp, &errors, " "),
```

File: src/wireless.c [hunk 12]
```diff
@@ -735,6 +853,42 @@ ni_wireless_wpa_map_pmf(ni_wireless_pmf_t p, int *ret)
...
+	const char *name;
+
+	if (!properties || !freq_list)
+		return FALSE;
+
+	if (freq_list->count == 0)
+		return TRUE;
+
+	if (!ni_wireless_frequency_list_expand(&frequencies, freq_list, &errors)) {
+		ni_error("Invalid values in frequency-list: %s", ni_stringbuf_join(&sb, &errors, ","));
```
Comment (mtomaschewski) [inline src/wireless.c:872]:
```suggestion
		ni_error("Invalid values in frequency-list: %s", ni_stringbuf_join(&sb, &errors, " "));
```

**Review by mtomaschewski — CHANGES_REQUESTED** [review]
LGTM, thanks!

I'd join the erroneous frequency values consistently with a space as well.

File: src/dbus-objects/wireless.c [hunk 13]
```diff
@@ -391,6 +391,23 @@ ni_objectmodel_get_wireless_request_net(const char *ifname, ni_wireless_network_
...
+						ifname);
```
Comment (mtomaschewski) [inline src/dbus-objects/wireless.c:403]: -> ifname context first, then (space separated) erroneous frequencies:
```suggestion
				dbus_set_error(error, DBUS_ERROR_INVALID_ARGS,
						"%s: invalid frequency-list value %s",
						ifname, ni_stringbuf_join(&tmp, &errors, " "));
```

**Review by mtomaschewski — COMMENTED** [review]
Ahm... almost overlooked it.

File: src/dbus-objects/wireless.c (hunk 13 above)
Comment (cfconrad) [reply src/dbus-objects/wireless.c:403]: good catch, thx!

**Review by mtomaschewski — APPROVED** [review]
LGTM, thanks!

**Pull request merged**